linkerd/core/src/main/scala/io/buoyant/linkerd/RequestActiveTracer.scala

```diff
@@ -0,0 +1,198 @@
+package io.buoyant.linkerd
+
+import com.twitter.finagle._
+import com.twitter.finagle.client.Transporter.EndpointAddr
+import com.twitter.finagle.http._
+import com.twitter.finagle.naming.buoyant.DstBindingFactory
+import com.twitter.util._
+import io.buoyant.namer.DelegateTree._
+import io.buoyant.namer.{DelegateTree, Delegator}
+import io.buoyant.router.RouterLabel
+import io.buoyant.router.RoutingFactory
+import io.buoyant.router.RoutingFactory.BaseDtab
+import io.buoyant.router.context.{DstBoundCtx, DstPathCtx}
+
+class RequestActiveTracer(
+  endpoint: EndpointAddr,
+  namer: DstBindingFactory.Namer,
+  dtab: BaseDtab,
+  routerLabel: String
+) extends SimpleFilter[Request, Response] {
+
+  private[this] case class Node(path: String, dentry: String, treeNode: DelegateTree[_]) {
+    override def toString: String = {
+      treeNode match {
+        case DelegateTree.Delegate(_, d, _) if d == Dentry.nop => s"$path"
+        case _ => s"$path ($dentry)"
+      }
+
+    }
+  }
+
+  private[this] val RequestTracerMaxDepthHeader = "l5d-max-depth"
+
+  private[this] def printEvaluatedRequest(
+    prependText: String,
+    serviceName: String,
+    clientName: String,
+    selectedAddress: String,
+    addresses: Option[Set[String]],
+    dtabResolution: List[String]
+  ): String = {
+    Seq(
+      prependText,
+      s"""
+         |--- Router: $routerLabel ---
+         |service name: $serviceName
+         |client name: $clientName
+         |addresses: [${addresses.getOrElse(Set.empty).mkString(", ")}]
+         |selected address: $selectedAddress
+         |dtab resolution:
+         |${dtabResolution.map("  " + _).mkString("\n")}
+    """.stripMargin
+    ).mkString("\n").trim.concat("\n")
+  }
+
+  /**
+   * formatDelegation prints out path the proxy used to resolve the client name.
+   * It 'walks' a DelegateTree of Name.Bound and finds a DelegateLeaf that matches
+   * the client path provided.
+   * @param dTree
+   * @param tree
+   * @param searchPath
+   * @return
+   */
+  private[this] def formatDelegation(
+    dTree: DelegateTree[Name.Bound],
+    tree: List[Node],
+    searchPath: Path
+  ): List[Node] = {
+    dTree match {
```

Comment:
this method is pretty dense.  breaking it up with comments or vals with intermediate values might help make it more readable.